src/workerd/api/streams/writable-sink-adapter.c++

```diff
@@ -0,0 +1,864 @@
+}
+
+bool WritableStreamSinkJsAdapter::isClosing() {
+  KJ_IF_SOME(active, state.tryGet<IoOwn<Active>>()) {
+    return active->closePending;
+  } else {
+    return false;
+  }
+}
+
+kj::Maybe<ssize_t> WritableStreamSinkJsAdapter::getDesiredSize() {
+  return state.tryGet<IoOwn<Active>>().map(
+      [](IoOwn<Active>& active) { return active->getDesiredSize(); });
+}
+
+jsg::Promise<void> WritableStreamSinkJsAdapter::write(jsg::Lock& js, const jsg::JsValue& value) {
+  KJ_SWITCH_ONEOF(state) {
+    KJ_CASE_ONEOF(activeOwn, IoOwn<Active>) {
+      // Deference the IoOwn once to get the active state.
+      auto& active = *activeOwn;
+
+      // If close is pending, we cannot accept any more writes.
+      if (active.closePending) {
+        auto exc = js.typeError("Write after close is not allowed");
+        // If we are configured to treat early errors as fatal, we need to
+        // abort the entire adapter.
+        return js.rejectedPromise<void>(exc);
+      }
+
+      // Ok, we are in a writable state, there are no pending closes.
+      // Let's process our data and write it!
+      auto& ioContext = IoContext::current();
+
+      // We know that a WritableStreamSink only accepts bytes, so we need to
+      // verify that the value is a source of bytes. We accept three possible
+      // types: ArrayBuffer, ArrayBufferView, and String. If it is a string,
+      // we convert it to UTF-8 bytes. Anything else is an error.
+      if (value.isArrayBufferView() || value.isArrayBuffer() || value.isSharedArrayBuffer()) {
+        // We can just wrap the value with a jsg::BufferSource and write it.
+        jsg::BufferSource source(js, value);
+        if (active.options.detachOnWrite && source.canDetach(js)) {
+          // Detach from the original ArrayBuffer...
+          // ... and re-wrap it with a new BufferSource that we own.
+          source = jsg::BufferSource(js, source.detach(js));
+        }
+
+        // Zero-length writes are a no-op.
+        if (source.size() == 0) {
+          return js.resolvedPromise();
+        }
+
+        active.bytesInFlight += source.size();
+        maybeSignalBackpressure(js);
+        // Enqueue the actual write operation into the write queue. We pass in
+        // two lambdas, one that does the actual write, and one that handles
+        // errors. If the write fails, we need to transition the adapter to the
+        // errored state. If the write succeeds, we need to decrement the
+        // bytesInFlight counter.
+        //
+        // The promise returned by enqueue is not the actual write promise but
+        // a branch forked off of it. We wrap that with a JS promise that waits
+        // for it to complete. Once it does, we check if we can release backpressure.
+        // This has to be done within in Isolate lock because we need to be able
+        // to resolve or reject the JS promises. If the write fails, we instead
+        // abort the backpressure state.
```

Comment:
Tiny stupid nit: add or remove newline to make the paragraph break more obvious, or all one paragraph